src/he/conn.c

```diff
@@ -416,12 +416,19 @@ he_return_code_t he_internal_send_message(he_conn_t *conn, uint8_t *message, uin
   // For now - this will just do a simple write and assume it works
   int res = wolfSSL_write(conn->wolf_ssl, message, (int)length);
 
-  // Note that we expect wolfSSL_write callback to ALWAYS succeed
-  // Any output buffering is done by the host app
-  if(res == 0) {
-    return HE_ERR_CONNECTION_WAS_CLOSED;
-  } else if(res < 0) {
-    return HE_ERR_SSL_ERROR;
+  if(res <= 0) {
+    int error = wolfSSL_get_error(conn->wolf_ssl, res);
+    switch(error) {
+      case SSL_ERROR_NONE:
+        // Previous API appeared to return an error code but no error actually occurred
+        return HE_SUCCESS;
+      case SSL_ERROR_WANT_WRITE:
+        // The underlying I/O is non-blocking and it could not satisfy the needs of wolfSSL_write to
+        // continue. Return HE_WANT_WRITE so that the caller can try again.
+        return HE_WANT_WRITE;
+      default:
+        return (res == 0) ? HE_ERR_CONNECTION_WAS_CLOSED : HE_ERR_SSL_ERROR;
```

Comment:
Because this now returns new error codes, do we have to update any documentation on this new behaviour?

Reply:
It's an internal function, no need to document it :-)

Reply:
Oh i meant the return codes for the public functions that this might affect. Sorry for being unclear.

Reply:
The `he_internal_send_message` is used by several public functions, so I don't think it worth the effort documenting every possible return values there. But maybe @expressvpn-pete-m has different opinion.